src/participant/prompts/promptHistory.ts

```diff
@@ -0,0 +1,198 @@
+      'emptyRequest',
+      'askToConnect',
+    ];
+
+    const responseType = (currentTurn.result as ChatResult)?.metadata?.intent;
+    if (responseTypesToSkip.includes(responseType)) {
+      // eslint-disable-next-line new-cap
+      return undefined;
+    }
+
+    // If the namespace is already known, skip including prompts asking for it.
+    if (responseType === 'askForNamespace' && namespaceIsKnown) {
+      // eslint-disable-next-line new-cap
+      return undefined;
+    }
+
+    for (const fragment of currentTurn.response) {
+      if (fragment instanceof vscode.ChatResponseMarkdownPart) {
+        message += fragment.value.value;
+
+        if (
+          (currentTurn.result as ChatResult)?.metadata?.intent ===
+          'askForNamespace'
+        ) {
+          // When the message is the assistant asking for part of a namespace,
+          // we only want to include the question asked, not the user's
+          // database and collection names in the history item.
+          break;
+        }
+      }
+    }
+
+    // eslint-disable-next-line new-cap
+    return vscode.LanguageModelChatMessage.Assistant(message);
+  }
+
+  private static _handleChatRequestTurn({
+    previousTurn,
+    currentTurn,
+    nextTurn,
+    connectionNames,
+    namespaceIsKnown,
+  }: {
+    previousTurn: vscode.ChatRequestTurn | vscode.ChatResponseTurn | undefined;
+    currentTurn: vscode.ChatRequestTurn;
+    nextTurn: vscode.ChatRequestTurn | vscode.ChatResponseTurn | undefined;
+    connectionNames: string[] | undefined;
+    namespaceIsKnown: boolean;
+  }): vscode.LanguageModelChatMessage | undefined {
+    if (previousTurn instanceof vscode.ChatResponseTurn) {
+      const responseIntent = (previousTurn.result as ChatResult).metadata
+        ?.intent;
+
+      if (responseIntent === 'askForNamespace' && namespaceIsKnown) {
+        // If the namespace is already known, skip responses to prompts asking for it.
+        return undefined;
+      }
+    }
+
+    if (
+      nextTurn instanceof vscode.ChatResponseTurn &&
+      nextTurn.result.errorDetails?.message === ParticipantErrorTypes.FILTERED
+    ) {
+      // If the response to this request led to a filtered error,
+      // we do not want to include it in the history
+      return undefined;
+    }
+
+    if (
+      currentTurn.prompt?.trim().length === 0 ||
+      connectionNames?.includes(currentTurn.prompt)
+    ) {
+      // When the message is empty or a connection name then we skip it.
+      // It's probably going to be the response to the connect step.
+      return undefined;
+    }
+
+    // eslint-disable-next-line new-cap
+    return vscode.LanguageModelChatMessage.User(currentTurn.prompt);
+  }
+
+  /** When passing the history to the model we only want contextual messages
+  to be passed. This function parses through the history and returns
+  the messages that are valuable to keep. */
+  static getFilteredHistory({
```

Comment:
this could also just be exported functions; the class just acts as namespacing for the time being so not sure if it makes sense